### PR TITLE
[ir] Represent Glow IR instructions lists using LLVM's intrusive lists

### PR DESCRIPTION
**Pull request by opti-mix**
This change solves a number of compilation performance issues, because instructions can be found in the list or removed from the list very quickly.

The change touches many of the loops iterating over the instruction lists.

While looking at the loops iterating over the Glow IR instructions, also take the chance to tighten their const correctness. This includes tightening the const correctness of some APIs used by those loops as well.

### CODE REVIEW

File: include/glow/IR/IR.h [hunk 1]
```diff
@@ -120,7 +126,10 @@ class Instruction : public Value {
   Instruction(const Instruction &I) = delete;
   Instruction &operator=(const Instruction &I) = delete;
 
-protected:
+  /// Destroy an instruction.
```
Comment (nadavrot) [inline include/glow/IR/IR.h]: Does this deallocate the instruction? Does it remove it from the list or do you need to remove it before?

File: include/glow/IR/IR.h [hunk 2]
```diff
@@ -140,12 +149,12 @@ class Instruction : public Value {
   /// Adds a new operand \p op at the end of the operand list.
   void pushOperand(Operand op);
 
-  Instruction(IRFunction *M, llvm::StringRef name, Kinded::Kind k, TypeRef Ty)
-      : Value(name, Ty, k), F_(M) {}
+  Instruction(llvm::StringRef name, Kinded::Kind k, TypeRef Ty)
```
Comment (nadavrot) [inline include/glow/IR/IR.h]: Why did you remove the parent method and put nullptr instead?

File: include/glow/IR/IR.h [hunk 3]
```diff
@@ -285,47 +327,29 @@ class IRFunction final {
   /// \returns the list of instructions.
   const InstListTy &getInstrs() const { return instrs_; }
 
+  /// getSublistAccess() - returns pointer to member of instruction list
```
Comment (nadavrot) [inline include/glow/IR/IR.h]: Why do you put the name of the function in the documentation? 

Also, Why is this called getSublistAccess ? Why not getInstructions() or something? What part is 'sub' ?

File: include/glow/IR/IRUtils.h [hunk 4]
```diff
@@ -123,13 +123,53 @@ class ValueUses {
   const_iterator end() const { return const_iterator{v_, 1}; }
 };
 
+/// A convenience iterator type that returns element pointers instead of element
```
Comment (nadavrot) [inline include/glow/IR/IRUtils.h]: I am okay with adding this as a temporary solution, but I think that we should remove this in the next commit and just change all uses.

File: include/glow/IR/IR.h (hunk 3 above)
Comment (opti-mix) [reply include/glow/IR/IR.h]: Copy&paste from LLVM code. Let's give it a more reasonable name.

File: include/glow/IR/IRUtils.h (hunk 4 above)
Comment (opti-mix) [reply include/glow/IR/IRUtils.h]: After a discussion with others, let's just use the standard ilist iterators returning references and update all the loop bodies.

File: include/glow/IR/IR.h (hunk 2 above)
Comment (opti-mix) [reply include/glow/IR/IR.h]: We were setting the parent in the constructor, before we even added the instruction to an IRFunction. Now the parent link is set when the instruction is added to the instruction list of a function for real.

File: include/glow/IR/IR.h (hunk 3 above)
Comment (opti-mix) [reply include/glow/IR/IR.h]: Fixed.

File: include/glow/IR/IRUtils.h (hunk 4 above)
Comment (opti-mix) [reply include/glow/IR/IRUtils.h]: Fixed.

**Comment by nadavrot** [comment]
Nice!
